eo-phi-normalizer/src/Language/EO/Phi/Syntax.hs

```diff
@@ -125,3 +227,242 @@ render d = rend 0 False (map ($ "") $ d []) ""
+
+-- | Encode 'String' as 'Bytes'.
+--
+-- >>> stringToBytes "Hello, world!"
+-- Bytes "48-65-6C-6C-6F-2C-20-77-6F-72-6C-64-21"
+--
+-- >>> stringToBytes "Привет, мир!"
+-- Bytes "D0-9F-D1-80-D0-B8-D0-B2-D0-B5-D1-82-2C-20-D0-BC-D0-B8-D1-80-21"
+--
+-- >>> stringToBytes  "hello, 大家!"
+-- Bytes "68-65-6C-6C-6F-2C-20-E5-A4-A7-E5-AE-B6-21"
+stringToBytes :: String -> Bytes
+stringToBytes s = bytestringToBytes $ Text.encodeUtf8 (Text.pack s)
+
+bytestringToBytes :: ByteString -> Bytes
+bytestringToBytes = Bytes . normalizeBytes . foldMap (padLeft 2 . (`showHex` "")) . ByteString.Strict.unpack
+
+bytesToByteString :: Bytes -> ByteString
+bytesToByteString (Bytes bytes) = ByteString.Strict.pack (go (filter (/= '-') bytes))
+ where
+  go [] = []
+  go (x : y : xs) = fst (head (readHex [x, y])) : go xs
+  go [_] = error "impossible: partial byte"
+
+-- | Decode 'String' from 'Bytes'.
+--
+-- >>> bytesToString "48-65-6C-6C-6F-2C-20-77-6F-72-6C-64-21"
+-- "Hello, world!"
+bytesToString :: Bytes -> String
+bytesToString = Text.unpack . Text.decodeUtf8 . bytesToByteString
+
+-- | Encode 'Double' as 'Bytes' following IEEE754.
+--
+-- Note: it is called "float" in EO, but it actually occupies 8 bytes so it corresponds to 'Double'.
+--
+-- >>> floatToBytes 0
+-- Bytes "00-00-00-00-00-00-00-00"
+--
+-- >>> floatToBytes (-0.1)
+-- Bytes "BF-B9-99-99-99-99-99-9A"
+--
+-- >>> floatToBytes (1/0)       -- Infinity
+-- Bytes "7F-F0-00-00-00-00-00-00"
+--
+-- >>> floatToBytes (asin 2) `elem` ["FF-F8-00-00-00-00-00-00", "7F-F8-00-00-00-00-00-00"]  -- sNaN or qNaN
+-- True
+floatToBytes :: Double -> Bytes
+floatToBytes f = Bytes $ normalizeBytes $ foldMap (padLeft 2 . (`showHex` "")) $ ByteString.Strict.unpack $ Serialize.encode f
+
+-- | Decode 'Double' from 'Bytes' following IEEE754.
+--
+-- >>> bytesToFloat "00-00-00-00-00-00-00-00"
+-- 0.0
+--
+-- >>> bytesToFloat "BF-B9-99-99-99-99-99-9A"
+-- -0.1
+--
+-- >>> bytesToFloat "7F-F0-00-00-00-00-00-00"
+-- Infinity
+--
+-- >>> bytesToFloat "FF-F8-00-00-00-00-00-00"
+-- NaN
+bytesToFloat :: Bytes -> Double
+bytesToFloat (Bytes bytes) =
+  case Serialize.decode $ ByteString.Strict.pack $ map (fst . head . readHex) $ words (map dashToSpace bytes) of
+    Left msg -> error msg
+    Right x -> x
+ where
+  dashToSpace '-' = ' '
+  dashToSpace c = c
+
+instance IsString Program where fromString = unsafeParseWith pProgram
+instance IsString Object where fromString = unsafeParseWith pObject
+instance IsString Binding where fromString = unsafeParseWith pBinding
+instance IsString Attribute where fromString = unsafeParseWith pAttribute
+instance IsString RuleAttribute where fromString = unsafeParseWith pRuleAttribute
+instance IsString PeeledObject where fromString = unsafeParseWith pPeeledObject
+instance IsString ObjectHead where fromString = unsafeParseWith pObjectHead
+
+instance IsString MetaId where fromString = unsafeParseWith pMetaId
+
+parseWith :: ([Token] -> Either String a) -> String -> Either String a
```

Comment:
@fizruk, this is an incorrect version of `parseWith`.

Reply:
Wow, did not notice when resolving conflicts, sorry.